File: models/war.py
"""
Модели данных для войн - аналог Java WarToSave и AttackData
"""
from dataclasses import dataclass
from typing import List, Dict, Optional


@dataclass
class AttackData:
    """Данные об атаке"""
    attacker_name: str
    stars: int
    destruction: float
    
    def __init__(self, attacker_name: str, stars: int, destruction: float):
        self.attacker_name = attacker_name
        self.stars = stars
        self.destruction = destruction


@dataclass
class WarToSave:
    """Данные войны для сохранения в БД"""
    end_time: str
    opponent_name: str
    team_size: int
    clan_stars: int
    opponent_stars: int
    clan_destruction: float
    opponent_destruction: float
    clan_attacks_used: int
    result: str
    is_cwl_war: bool
    total_violations: int
    attacks_by_member: Dict[str, List[Dict]] = None
    
    def __init__(self, end_time: str, opponent_name: str, team_size: int, 
                 clan_stars: int, opponent_stars: int, clan_destruction: float,
                 opponent_destruction: float, clan_attacks_used: int, result: str,
                 is_cwl_war: bool, total_violations: int, attacks_by_member: Dict = None):
        self.end_time = end_time
        self.opponent_name = opponent_name
        self.team_size = team_size
        self.clan_stars = clan_stars
        self.opponent_stars = opponent_stars
        self.clan_destruction = clan_destruction
        self.opponent_destruction = opponent_destruction
        self.clan_attacks_used = clan_attacks_used
        self.result = result
        self.is_cwl_war = is_cwl_war
        self.total_violations = total_violations
        self.attacks_by_member = attacks_by_member or {}


@dataclass
class Player:
    """Модель игрока из COC API"""
    tag: str
    name: str
    town_hall_level: int
    trophies: int
    clan: Optional[Dict] = None
    
    
@dataclass
class Clan:
    """Модель клана из COC API"""
    tag: str
    name: str
    description: str
    location: Optional[Dict] = None
    members: List[Dict] = None
    war_wins: int = 0
    war_losses: int = 0
    war_ties: int = 0

File: internal/models/war.go
package models

import "time"

// WarToSave represents a clan war to be saved
type WarToSave struct {
	EndTime                       string    `json:"end_time"`
	OpponentName                  string    `json:"opponent_name"`
	TeamSize                      int       `json:"team_size"`
	ClanStars                     int       `json:"clan_stars"`
	OpponentStars                 int       `json:"opponent_stars"`
	ClanDestructionPercentage     float64   `json:"clan_destruction_percentage"`
	OpponentDestructionPercentage float64   `json:"opponent_destruction_percentage"`
	ClanAttacksUsed               int       `json:"clan_attacks_used"`
	Result                        string    `json:"result"` // "win", "lose", "tie"
	IsCWL                         bool      `json:"is_cwl"`
	TotalViolations               int       `json:"total_violations"`
	Attacks                       []AttackData `json:"attacks"`
}

// AttackData represents an attack in a clan war
type AttackData struct {
	AttackerName  string  `json:"attacker_name"`
	DefenderTag   string  `json:"defender_tag"`
	Stars         int     `json:"stars"`
	Destruction   float64 `json:"destruction"`
	Order         int     `json:"order"`
	Timestamp     int     `json:"timestamp"`
	IsViolation   int     `json:"is_violation"`
}

// NewWarToSave creates a new WarToSave instance with all parameters
func NewWarToSave(
	endTime string,
	opponentName string,
	teamSize int,
	clanStars int,
	opponentStars int,
	clanDestruction float64,
	opponentDestruction float64,
	clanAttacksUsed int,
	result string,
	isCWLWar bool,
	totalViolations int,
) *WarToSave {
	return &WarToSave{
		EndTime:                       endTime,
		OpponentName:                  opponentName,
		TeamSize:                      teamSize,
		ClanStars:                     clanStars,
		OpponentStars:                 opponentStars,
		ClanDestructionPercentage:     clanDestruction,
		OpponentDestructionPercentage: opponentDestruction,
		ClanAttacksUsed:               clanAttacksUsed,
		Result:                        result,
		IsCWL:                         isCWLWar,
		TotalViolations:               totalViolations,
	}
}

// NewAttackData creates a new AttackData instance
func NewAttackData(attackerName, defenderTag string) *AttackData {
	return &AttackData{
		AttackerName: attackerName,
		DefenderTag:  defenderTag,
		IsViolation:  0,
	}
}
